### ISSUE TITLE
Trouble getting the redirect_uri into the challenge in a new Azure B2C app

### ISSUE THREAD

**Issue by GuardRex** [aspnet/Security#481]
I tried asking in the [AzureAdQuickStarts/B2C-WebApp-OpenIdConnect-DotNet](https://github.com/AzureADQuickStarts/B2C-WebApp-OpenIdConnect-DotNet) repo, but it doesn't look like the issues there are being monitored/answered.

I got into Azure B2C today ... and with my fav framework, Core CLR (beta7) and `Microsoft.AspNet.Mvc.Core: 6.0.0-beta7`.

I took a look at the sample we have, but it uses ...
```
HttpContext.GetOwinContext().Authentication.Challenge(
    new AuthenticationProperties (
        new Dictionary<string, string> 
        { 
            {Startup.PolicyKey, Startup.SignInPolicyId}
        })
        { 
            RedirectUri = "/", 
        }, OpenIdConnectAuthenticationDefaults.AuthenticationType
);
```
which is available from one of the packages in the sample project that isn't Core CLR-friendly.

I was successful at getting it working based on the earlier Core CLR app that I had created for vanilla AAD security ... but there is one problem. The original code from my earlier AAD app uses `ChallengeResult` ...
```
return new ChallengeResult(
    OpenIdConnectAuthenticationDefaults.AuthenticationScheme, 
    new AuthenticationProperties { RedirectUri = "/" }
);
```
... and I tried several formats of the statement to include the B2C policy designation without luck. My requests using that approach result in "There is no Redirect Url" exceptions." from AAD (B2C-type directory). It's true. Manually checking the Url produced, I can see that the `redirect_url` was not being put into the Url.

I was able to manually create the link (and including the policy):
```
Context.Response.Redirect("https://login.microsoftonline.com/XXXXXX.onmicrosoft.com/oauth2/
v2.0/authorize?p=B2C_1_SigninPolicy
&client_id=XXXXXXXXXXXXX
&response_type=id_token
&scope=openid
&response_mode=form_post
&redirect_uri=https%3a%2f%2fwww%2XXXXX%2eXXXXX%2ecom%3a8008%2f");
```
That works; but given that I'm not using the nonce here, I had to disable the nonce in the options:
```
options.ProtocolValidator.RequireNonce = false;
```
... not good! Definitely _**not**_ what I want to do.

Is it possible to use the `ChallengeResult` code to get the `redirect_url` in the request Uri? ... how should that be formatted?

If it's not possible at this time because the library isn't ready for Core CLR+Azure B2C yet, can I go on creating the link manually and add the nonce to it manually? ... If possible, how can I get a valid nonce for the link?

**Comment by Tratcher** [comment]
RedirectUri on the challenge is where you would like to have the user returned to after the all of the authentication is complete. redirect_url is the local url where the OIDC middleware will be listening for requests. You can set it via OpenIdConnectOptions.RedirectUri https://github.com/aspnet/Security/blob/dev/src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectOptions.cs#L163

**Comment by GuardRex** [comment]
Yes, but I should be able to set it in the challenge as well (I thought) so that challenges issued in different places in the app can return the user to the view where the challenge was issued or to some other destination on return. Whatever the case on setting the `redirect_uri` in a challenge, I did set it as you suggested in the `OpenIdConnectAuthentation` options ...
```
options.RedirectUri = "https://www.XXXXX.com/";
```
... which does work and solved that problem in a generic way. AAD B2C isn't complaining about its absence any longer, and AAD B2C returns to the app.

However, issues are still present trying to get the challenges to be phrased properly.

One problem is on logout. I got this straight [with the team's help (Security #356)](https://github.com/aspnet/Security/issues/356) for my earlier app with ...
```
[HttpGet]
public async Task LogOut()
{
    await Context.Authentication.SignOutAsync(OpenIdConnectAuthenticationDefaults.AuthenticationScheme);
    await Context.Authentication.SignOutAsync(CookieAuthenticationDefaults.AuthenticationScheme);
}
```
... which works in my earlier AAD app. In this AD B2C app, however, it throws the user up to AD B2C like the `ProfilePolicyId` has been provided:

![capture](https://cloud.githubusercontent.com/assets/1622880/9974643/5256ddfc-5e5f-11e5-9dfe-f1f16fa6f88c.PNG)

... and that's also what happens when I try to use this for login code ...
```
return new ChallengeResult(
    OpenIdConnectAuthenticationDefaults.AuthenticationScheme,
    new AuthenticationProperties(
        new Dictionary<string, string> {
            { Startup.PolicyKey, Startup.SignInPolicyId }
        }
    )
);
```
It looks like to me that `ChallengeResult` in the `Mvc.Core` namespace isn't capable of composing Url's correctly for AAD B2C ... or I can't figure it out anyway. I can wait until that comes along later. I'm fine with my workaround of manually using a static address for now. Is there a way that I can generate a valid nonce for my manually-provided Uri challenge?
```
Context.Response.Redirect("https://login.microsoftonline.com/XXXXXX.onmicrosoft.com/oauth2/
v2.0/authorize?p=B2C_1_SigninPolicy
&client_id=XXXXXXXXXXXXX
&response_type=id_token
&scope=openid
&response_mode=form_post
&redirect_uri=https%3a%2f%2fwww%2XXXXX%2eXXXXX%2ecom%3a8008%2f"
&nonce=_______________________________________
);
```
